tests/debugging/function/test_discovery.py

```diff
@@ -157,3 +158,27 @@ def f():
     assert f is _undecorate(f, name, path)
 
     assert _undecorate(_undecorate, name, path) is _undecorate
+
+
+def test_discovery_class_decoration():
+    class Decorator:
+        def __init__(self, f):
+            self.f = f
+
+    @Decorator
+    def f():
+        pass
+
+    code = _undecorate(f, name="f", path=abspath(__file__)).__code__
+    assert code.co_name == "f"
+    assert code.co_filename == abspath(__file__)
+
+
+def test_discovery_wrapped_decoration():
```

Comment:
should we also test for partial?

Reply:
`wraps` uses `partial` for its implementation
```
def wraps(wrapped,
          assigned = WRAPPER_ASSIGNMENTS,
          updated = WRAPPER_UPDATES):
    """Decorator factory to apply update_wrapper() to a wrapper function

       Returns a decorator that invokes update_wrapper() with the decorated
       function as the wrapper argument and the arguments to wraps() as the
       remaining arguments. Default arguments are as for update_wrapper().
       This is a convenience function to simplify applying partial() to
       update_wrapper().
    """
    return partial(update_wrapper, wrapped=wrapped,
                   assigned=assigned, updated=updated)
```

Reply:
I see. this is cool. while it two birds in one stone situation here, I think we shouldn't depend on wraps implementation and have it explicit if wraps impl changed in the future. WDYT?

Reply:
I think it's hard for this implementation to change in a way that won't use `partial` in the future, but we can add a dedicated test case anyway.